### ISSUE TITLE
Question about data format

### ISSUE THREAD

**Issue by Zephyr1022** [GanjinZero/Triaffine-nested-ner#4]
I am trying to apply the model to other data corpus. I was curious that what is your input data format. How could I preprocess the data and then pass to the model? Thank you.

**Comment by GanjinZero** [comment]
Data format can found here **https://github.com/GanjinZero/Triaffine-nested-ner/tree/main/data/samples/ace04**.
You should contain entity, tokens, ltokens (left sentence in same paragraph), rtokens (right sentence in same paragraph), pos information for each sentence. ltokens and rtokens can be a [], if you don't have context information. pos are obtained by **nltk.pos_tag**.
For new data, you need to modify **_get_type** in **data_util.py** and fix **version** related code anywhere in this repo (version contains dataset meta information).
And you need to use **word_embed.py** to generate word embeddings for your dataset.
```python
generate_vocab_embed(your_dataset, 'cc.en.300.bin', 'fasttext')
```
There may be more details for using on other dataset, please feel free to ask me if you have any other problems.

**Comment by Zephyr1022** [comment]
> Data format can found here **https://github.com/GanjinZero/Triaffine-nested-ner/tree/main/data/samples/ace04**. You should contain entity, tokens, ltokens (left sentence in same paragraph), rtokens (right sentence in same paragraph), pos information for each sentence. ltokens and rtokens can be a [], if you don't have context information. pos are obtained by **nltk.pos_tag**. For new data, you need to modify **_get_type** in **data_util.py** and fix **version** related code anywhere in this repo (version contains dataset meta information). And you need to use **word_embed.py** to generate word embeddings for your dataset.
> 
> ```python
> generate_vocab_embed(your_dataset, 'cc.en.300.bin', 'fasttext')
> ```
> 
> There may be more details for using on other dataset, please feel free to ask me if you have any other problems.

Thank you so much for your reply.  

I am a little confused about the entities' start and end indexes. For example, the first entity type: GPE start 4 and end 5 are mapped to the token "Shanghai"? And in the input data, what symbols are you using to split the different files? The example from ace04 seems like fews senteces from one document. I have multiple this kind documents. Do you use [] to concatenate different documents, like [doc1] [doc2] ...? 

One more question, when I change version to my dataset name. It gives this following error back: main.py: error: argument --version: invalid choice: 'sdoh' (choose from 'ace05', 'ace04', 'genia91', 'kbp'). How could I fix this issue? Thank you.
 {
        "entities": [
            {
                "type": "GPE",
                "start": 4,
                "end": 5
            },
            {
                "type": "ORG",
                "start": 0,
                "end": 3
            },
            {
                "type": "PER",
                "start": 10,
                "end": 11
            },
            {
                "type": "PER",
                "start": 10,
                "end": 13
            }
        ],
        "relations": [],
        "tokens": [
            "Xinhua",
            "News",
            "Agency",
            ",",
            "Shanghai",
            ",",
            "August",
            "31st",
            ",",
            "by",
            "reporter",
            "Jierong",
            "Zhou"
        ],

**Comment by GanjinZero** [comment]
(1) GPE start 4 and end 5  = [4,5)
(2) I remember ace04 save each document in different files, so documents are segmented naturally. I do not concatenate documents.
(3) Add your dataset config in **train_parser.py**:
```python
    parser.add_argument("--version", type=str, default="genia",
                        choices=["ace05", "ace04", "genia91", "kbp", "sdoh"],
                        help="Dataset version.")
```
You need to change everything related to args.version.

**Comment by Zephyr1022** [comment]
Thank you very much for the explanation.

My current data format is shown below. Each line of the input file contains one document. I would like to change this to fit your input format. Am I supposed to separate each sentence into a different {} ?  

{"doc_key": "./1234.txt", "sentences": [["SOCIAL", "HISTORY", ":", "Lives", "with", "his", "caring", "and", "devoted", "parents", "at", "home", "."], ["Enjoys", "movies", "and", "computers", "."], ["No", "history", "of", "alcohol", ",", "tobacco", "or", "drug", "use", "."]], "ner": [[[3, 3, "StatusTime"], [4, 9, "TypeLiving"]], [], [[18, 19, "StatusTime"]]], "relations": [[[3, 3, 3, 3, "LivingStatus-Status"], [3, 3, 4, 9, "LivingStatus-Type"]], [], [[23, 23, 18, 19, "Tobacco-Status"], [25, 26, 18, 19, "Drug-Status"], [21, 21, 18, 19, "Alcohol-Status"]]]}
{"doc_key": "./2345.txt", "sentences": [["Social", "History", ":", "Lives", "in", "group", "home", "in", "[", "*", "*", "Hospital1", "3494", "*", "*", "]", ",", "dependent", "on", "aides", "for", "ADLs", "."], ["Per", "OMR", ",", "lifetime", "non", "-", "smoker", ",", "no", "h", "/", "o", "heavy", "etoh", "use", "."], ["Has", "brother", "[", "*", "*", "Name", "(", "NI", ")", "*", "*", "]", "who", "is", "his", "HCP", "."]], "ner": [[[3, 3, "StatusTime"], [4, 15, "TypeLiving"]], [[31, 34, "StatusTime"], [26, 29, "StatusTime"]], []], "relations": [[[3, 3, 3, 3, "LivingStatus-Status"], [3, 3, 4, 15, "LivingStatus-Type"]], [[35, 37, 31, 34, "Alcohol-Status"], [26, 29, 26, 29, "Tobacco-Status"]], []]}

And I was also curious about where I could find the prediction file? Thank you.

**Comment by GanjinZero** [comment]
Each sentence is a dict {}.
The whole file is like [{}, {}, {}, ...].

Predictions will be obtained each epoch in your output folder.

**Comment by Zephyr1022** [comment]
Cool, Thank you so much. That's very helpful. I have a last question that do you have any trick for setting hyperparameters? I tried some hyperparameters combinations, but it is still slow to learn on my dataset.

**Comment by GanjinZero** [comment]
You can try bigger learning rate like 5e-5, it does take some epochs to learn for every dataset (50 epochs for ACE04 and ACE05).

**Comment by GanjinZero** [comment]
You can also set encoder_learning_rate bigger, but I have not tried on it.

**Comment by Zephyr1022** [comment]
Got it. Thank you so much.
